lib/collection/src/shards/transfer/shard_transfer.rs

```diff
@@ -790,41 +786,37 @@ where
                 .await
             };
 
-            result = cancel::future::cancel_on_token(cancel.clone(), future)
-                .await
-                .map_err(Into::into)
-                .and_then(|res| res);
-
-            match &result {
-                Ok(()) => break,
+            result = cancel::future::cancel_on_token(cancel.clone(), future).await;
 
-                // TODO: Should we break early if shard transfer is cancelled, or should we handle it as regular error?
-                Err(CollectionError::Cancelled { .. }) => break,
-
-                Err(err) => {
-                    log::error!(
-                        "Failed to transfer shard {} -> {}: {err}",
-                        transfer.shard_id,
-                        transfer.to,
-                    );
+            // If `transfer_shard` returned error...
+            if let Ok(Err(err)) = &result {
+                log::error!(
+                    "Failed to transfer shard {collection_id}:{} -> {}: {err}",
+                    transfer.shard_id,
+                    transfer.to,
+                );
+            }
 
-                    // Revert queue proxy if we still have any to prepare for the next attempt
-                    if let Some(shard) = shards_holder.read().await.get_shard(&transfer.shard_id) {
-                        shard.revert_queue_proxy_local().await;
-                    }
+            // If `transfer_shard` returned error or task was cancelled...
+            if matches!(result, Ok(Err(_)) | Err(_)) {
+                // Revert queue proxy if we still have any to prepare for the next attempt
+                if let Some(shard) = shards_holder.read().await.get_shard(&transfer.shard_id) {
+                    shard.revert_queue_proxy_local().await;
                 }
             }
+
+            // If `transfer_shard` returned success or task was cancelled...
+            if !matches!(result, Ok(Err(_))) {
```

Comment:
This may be more readable, though...

```suggestion
            if matches!(result, Ok(Ok(())) | Err(_)) {
```

Reply:
Yeah lets do this one. I was a bit confused by the negation.